kernel/tty/src/discipline.rs

```diff
@@ -0,0 +1,266 @@
+            *input_buf = Vec::new();
+        }
+    }
+
+    pub(crate) fn process_input_byte(
+        &self,
+        byte: u8,
+        master: &Channel,
+        slave: &Channel,
+    ) -> Result<()> {
+        const ERASE: u8 = 0x7f; // DEL (backspace key)
+        const WERASE: u8 = 0x17; // ^W
+
+        const INTERRUPT: u8 = 0x3;
+        const SUSPEND: u8 = 0x1a;
+
+        match byte {
+            INTERRUPT => {
+                let _ = self.manager.send(Event::CtrlC);
+                self.clear_input_buf();
+                return Ok(());
+            }
+            SUSPEND => {
+                let _ = self.manager.send(Event::CtrlZ);
+                self.clear_input_buf();
+                return Ok(());
+            }
+            _ => {}
+        }
+
+        // TODO: EOF and EOL
+        // TODO: UTF-8?
+        if self.echo.load(Ordering::SeqCst) {
+            match (byte, &*self.canonical.lock().unwrap()) {
+                // TODO: Also pass-through START and STOP characters
+                (b'\t' | b'\n', _) => {
+                    master.send(byte)?;
+                }
+                (ERASE, Some(input_buf)) => {
+                    if !input_buf.is_empty() {
+                        master.send_buf([0x8, b' ', 0x8])?
+                    }
+                }
+                (WERASE, Some(input_buf)) => {
+                    if !input_buf.is_empty() {
+                        // TODO: Cache offset. Currently we're calculating it twice: once here,
+                        // and once if canonical mode is enabled.
+                        let offset = werase(input_buf);
+                        for _ in 0..offset {
+                            master.send_buf([0x8, b' ', 0x8])?;
+                        }
+                    }
+                }
+                (0..=0x1f, _) => {
+                    master.send_buf([b'^', byte + 0x40])?;
+                }
+                _ => {
+                    master.send(byte)?;
+                }
+            }
+        }
+
+        if let Some(ref mut input_buf) = *self.canonical.lock().unwrap() {
+            match byte {
+                b'\r' | b'\n' => {
+                    slave.send_buf(core::mem::take(input_buf))?;
+                    slave.send(b'\n')?;
+                }
+                ERASE => {
+                    input_buf.pop();
+                }
+                WERASE => {
+                    for _ in 0..werase(input_buf) {
+                        input_buf.pop();
+                    }
+                }
+                _ => input_buf.push(byte),
+            }
+        } else {
+            slave.send(byte)?;
+        }
+        Ok(())
+    }
+
+    pub(crate) fn process_input_buf(
+        &self,
+        buf: &[u8],
+        master: &Channel,
+        slave: &Channel,
+    ) -> Result<()> {
+        for byte in buf {
+            // TODO: This locks internal channels on every byte.
```

Comment:
the serial port impl already attempts to read multiple bytes (more specifically, `u16` input events) on every interrupt and send larger batches (up to 256 `u16`s for a total max of 512 bytes) per msg, so you could take a similar batching approach here